cmd/kube-dns/SPEC.md

```diff
@@ -0,0 +1,140 @@
+    - Answer Format:
+      - `_<port-x>._<proto-x>.<service>.<ns>.svc.<zone>. <ttl> IN SRV <weight> <priority> <port-number> <service>.<ns>.svc.<zone>.`
+    - Answer Example:
+```
+      _https._tcp.headless.default.svc.cluster.local.  4 IN SRV 10 100 443 my-pet-1.headless.default.svc.cluster.local.
+      _http._tcp.headless.default.svc.cluster.local.   4 IN SRV 10 100 80  my-pet-1.headless.default.svc.cluster.local.
+      _syslog._udp.headless.default.svc.cluster.local. 4 IN SRV 10 100 514 my-pet-1.headless.default.svc.cluster.local.
+      _https._tcp.headless.default.svc.cluster.local.  4 IN SRV 10 100 443 my-pet-2.headless.default.svc.cluster.local.
+      _http._tcp.headless.default.svc.cluster.local.   4 IN SRV 10 100 80  my-pet-2.headless.default.svc.cluster.local.
+      _syslog._udp.headless.default.svc.cluster.local. 4 IN SRV 10 100 514 my-pet-2.headless.default.svc.cluster.local.
+      _https._tcp.headless.default.svc.cluster.local.  4 IN SRV 10 100 443 438934893.headless.default.svc.cluster.local.
+      _http._tcp.headless.default.svc.cluster.local.   4 IN SRV 10 100 80  438934893.headless.default.svc.cluster.local.
+      _syslog._udp.headless.default.svc.cluster.local. 4 IN SRV 10 100 514 438934893.headless.default.svc.cluster.local.
+```
+- **Question:** `<reverse-ip-name>. IN PTR`
+  - If a Service with ClusterIP equal to the IP of `<reverse-ip-name>` exists, a PTR record with `<service>.<ns>.svc.<zone>.` will be returned.
+    - Answer Format:
+      - `<reverse-ip-name>. <ttl> IN PTR <service>.<ns>.svc.<zone>.`
+    - Answer Example:
+      - `1.0.3.10.in-addr.arpa. 14 IN PTR kubernetes.default.svc.cluster.local.`
+  - If a _ready_ endpoint exists with IP equal to the IP of `<reverse-ip-name>` exists, and the endpoint has a non-empty _hostname_ with value `<hostname>`, then a PTR record with `<hostname>.<service>.<ns>.svc.<zone>` will be returned.
+    - Answer Format:
+      - `<reverse-ip-name>. <ttl> IN PTR <hostname>.<service>.<ns>.svc.<zone>.`
+    - Answer Example:
+      - `100.0.3.10.in-addr.arpa. 14 IN PTR my-pet.my-service.default.svc.cluster.local.`
+  - Otherwise, the result is `NXDOMAIN`.
+- **Question:** `dns-version.<zone>. IN TXT`
+  - Returns the semantic versioning of the DNS-schema in use in this cluster.
+    - Answer Format:
+      - `dns-version.<zone>. <ttl> IN TXT <schema-version>`
+    - Answer Example:
+      - `dns-version.cluster.local. 28800 IN TXT "1.0.0"`
+
+## Schema Extensions
```

Comment:
We should probably go into more detail about extensions cannot overlap their domain names with the names given above? Otherwise, you will have incompatible extensions.
Some restrictions off the top of my head:
- For a given name, cannot have additional records be present for types specified by this doc (e.g. you cannot stick more A records where A records already exist).
- For a given class of names (e.g. <service>.<ns>.svc.<cluster>), extensions cannot have records that intersect with that set.